### PR TITLE
docs: write patterns guide and example.

### PR DESCRIPTION
**Pull request by thruflo**
See https://deploy-preview-2021--electric-next.netlify.app/docs/guides/writes

(when following a GitHub link to the example you'll need to `s/main/thruflo/writes-guide`)

### CODE REVIEW

File: examples/write-patterns/patterns/3-combine-on-read/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+- building local-first software
+- interactive SaaS applications
+- collaboration and authoring software
+
+## Drawbacks
+
+Combining data on-read makes local reads slightly slower.
+
+Using a local embedded database adds a relatively-heavy dependency to your app. This impacts build/bundle size, initialization speed and memory use. The shadow table and trigger machinery complicate your client side schema definition.
```
Comment (KyleAMathews) [inline examples/write-patterns/patterns/3-combine-on-read/README.md:35]: Perhaps clarify that this is a specific downside of pglite and that the same pattern could be implemented in other ways e.g. here with [valtio](https://github.com/pmndrs/valtio):

```ts
import { proxy, useSnapshot } from 'valtio'
import { useShape } from '@electric-sql/react'

interface Todo {
  id: number
  title: string
  completed: boolean
}

const optimisticStore = proxy({
  updates: new Map<number, Partial<Todo>>(),
  errors: new Map<number, string>()
})

const useTodos = () => {
  const { data: electricTodos, isLoading, error: electricError } = useShape<Todo>({
    url: 'http://localhost:3000/v1/shape',
    params: {
      table: 'todos',
      columns: ['id', 'title', 'completed']
    }
  })

  const { updates, errors } = useSnapshot(optimisticStore)

  const combinedTodos = isLoading ? [] : electricTodos.map(todo => ({
    ...todo,
    ...updates.get(todo.id),
    error: errors.get(todo.id)
  }))

  return { 
    todos: combinedTodos, 
    isLoading, 
    error: electricError 
  }
}

const updateTodo = (id: number, update: Partial<Todo>) => {
  optimisticStore.errors.delete(id)
  optimisticStore.updates.set(id, {
    ...optimisticStore.updates.get(id),
    ...update
  })

  fetch(`/api/todos/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(update)
  })
    .then(() => {
      optimisticStore.updates.delete(id)
    })
    .catch((err) => {
      optimisticStore.updates.delete(id)
      optimisticStore.errors.set(id, err.message)
    })
}

export { useTodos, updateTodo, optimisticStore }
```

Comment (KyleAMathews) [reply examples/write-patterns/patterns/3-combine-on-read/README.md:35]: There's of course lots of reasons to choose pglite but the drawbacks here are specific to pglite and not to the pattern in general.

The pattern is basically just pull the optimistic state maintenance out of the component and into a global singleton that any component can read from.

Comment (thruflo) [reply examples/write-patterns/patterns/3-combine-on-read/README.md:35]: Yup, maybe it is simpler to have this pattern in JS. Certainly if as is, I should clarify in the readme that pure JS patterns are equally valid and that the weaknesses mentioned are specific to the implementation.

I'll have a pass on it.

File: website/docs/guides/writes.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    1. a [comprehensive PGlite example](#); and
+    1. an [example using the external LiveStore framework](#)
+
+### Online writes
+
+no local state
+
+weaknesses:
+
+- takes a while
```
Comment (KyleAMathews) [inline website/docs/guides/writes.md:60]:
```suggestion
- the UI doesn't update until the server responds
```

File: website/docs/guides/writes.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+- need to be online
+
+### Optimistic state
+
+local optimistic state without persistence
+
+1. (2) optimistic state without persistence
+- useOptimistic react hook
```
Comment (KyleAMathews) [inline website/docs/guides/writes.md:68]: also worth mentioning other popular tools like tanstack/query, react-router (https://reactrouter.com/start/framework/pending-ui), Solid (https://docs.solidjs.com/solid-router/reference/data-apis/action) just to show it's a very common pattern across UI tooling that you can combine with Electric

File: packages/typescript-client/src/types.ts [hunk 4]
```diff
@@ -23,14 +23,16 @@ interface Header {
...
+
 export type ControlMessage = {
   headers: Header & { control: `up-to-date` | `must-refetch` }
 }
 
 export type ChangeMessage<T extends Row<unknown> = Row> = {
   key: string
   value: T
-  headers: Header & { operation: `insert` | `update` | `delete` }
+  headers: Header & { operation: Operation }
```
[the remark this answers is not in the document]
Comment (thruflo) [reply packages/typescript-client/src/types.ts:35]: Just drying the code up now that the `Operation` type exists. It's literally the same value.

**Review by KyleAMathews — APPROVED** [review]
Love it! :shipit:

**Pull request merged**